Nojira | More robust URL processing; prevent empty redirect entries causing build error

yvonnetangsu opened the pull request:
# READY FOR REVIEW

# Summary
- Use `sbStripSlugUrl` utility to remove prefix from internal links, but add more robust href processing

# Review By (Date)
- Soon

# Criticality
- 5

# Review Tasks

## Setup tasks and/or behavior to test

1. Go to the preview
2. Click some links in the masthead
3. Click some story cards
4. Click on a taxonomy chip to go to a story list page
5. Click on some story cards there
6. Go to the 404 page and check that CTA goes back to the homepage
7. Clicky clicky some more
8. Check that the sitemap still works the same https://deploy-preview-362--giving-campaign.netlify.app/sitemap.xml
9. Inspect the metadata on a couple pages and check that canonical URL etc still works properly
10. Look at code

yvonnetangsu commented on components/Storyblok/SbStoryCard.tsx:
```diff
@@ -85,7 +85,7 @@ export const SbStoryCard = ({
       tabColor={paletteAccentColors[value || storyTabColorValue]}
       link={link}
-      href={`/${full_slug}`}
+      href={full_slug}
```
We remove all leading slash before we check for the prefix in the updated sbStripSlugUrl utility so we don't need to add a leading slash here anymore

sherakama commented on utilities/sbStripSlugUrl.ts:
```diff
@@ -31,7 +31,25 @@ export const sbStripSlugURL = (url:string): string => {
+
+  /**
+   * If the URL is just the prefix, return a slash.
+   * Occasionally the cached_url for the momentum homepage still shows up as
+   * 'home' so adding the last check just in case.
```
Where is `home` coming from? The `real_path`?

yvonnetangsu replied:
I think it's somewhere in the cache 😬 It only happens sometimes in Storyblok (they have some weird caching issue when on their site occasionally). When I looked at the draft JSON, "home" was not the slug, but I've seen it rendered a few times in the editor so just putting it here just in case.

sherakama replied:
Huh. Ok. Thanks for explaining.

sherakama commented on components/Cta/CtaNextLink.tsx:
```diff
@@ -32,27 +33,14 @@ export const CtaNextLink = React.forwardRef<HTMLAnchorElement, CtaNextLinkProps>
-  }
-
-  const strippedHref = `/${cleanParts.join('/')}`;
+  const strippedHref = sbStripSlugURL(href);
+  const processedHref = strippedHref.startsWith('/') ? strippedHref : `/${strippedHref}`;
```
I think we should change `sbStripSlugURL` to always return with the leading slash instead of trying to handle this condition outside of it everywhere else.

yvonnetangsu replied:
Sure I can do that!

yvonnetangsu commented:
@sherakama I just found something weird, but maybe you know this already? I guess the `cached_url` really is cached, but for some reason I remember it changing when a story is moved in the past 🤔 

I did an experiment and went to a test page I added a long time ago before we did the directory split. There's an existing CTA to the changemaker page (on the left). I added a new one on the right that links to the same changemaker page. See how they have different `cached_url` - the old one is still changemaker but the new one has `cached_url` momentum/changemaker 👀  I've already opened up the old CTA on the left to jiggle it to make sure it has updated itself before I saved and looked at the draft JSON.

![Screenshot 2024-10-17 at 10 48 40 AM](https://github.com/user-attachments/assets/4266e722-2f0d-4db8-a878-708322f6c591)
![Screenshot 2024-10-17 at 10 48 33 AM](https://github.com/user-attachments/assets/2d34a1e1-18d3-4256-8b7a-92855b3a8562)

yvonnetangsu commented on utilities/data/getStoryblokRedirects.mjs:
```diff
@@ -100,12 +100,13 @@ export const getStoryblokRedirects = async () => {
 
-  const redirects = stories.map(entry => ({
+  const redirects = stories
+  .filter(entry => entry.content.from && entry.content.to) // Filter out entries without "from" or "to"
+  .map(entry => ({
     source: sanitizeSourcePath(entry.content.from),
     has: extractQueryParameters(entry.content.from),
     destination: entry.content.to,
-    statusCode: sanitizeRedirectCode(entry.content.statusCode),
+    statusCode: sanitizeRedirectCode(entry.content.statusCode || 301),
```
@sherakama Something interesting about creating redirects - Even though the `from` and `to` fields in Storyblok are required, and technically you can't save a redirect without entering something in those fields, the entry still exists without any values in those fields the moment it's created, so the empty test entry that you created in the demo today actually got "saved" and held up the Netlify build by causing an error 😆 Added this to filter out any entries without `to` and `from` fields, and add a fall back status code 301 here.

![Screenshot 2024-10-17 at 2 01 33 PM](https://github.com/user-attachments/assets/63d0c346-48d8-468d-bd35-7959d135896c)

sherakama replied:
Oh that is kinda ridiculous. More Storyblok fun!

yvonnetangsu replied:
I added an empty redirect to test and this PR preview built alright after I pushed up the commit to filter out those entries. I have now deleted that draft empty entry because it blocks my other PR preview 😂

yvonnetangsu commented on components/Storyblok/SbStoryListNav/SbStoryListNav.tsx:
```diff
@@ -26,21 +26,23 @@ type SbStoryListNavType = {
-      {links.map((link) => {
-        const cached_url = link.link?.cached_url ? sbStripSlugURL(link.link.cached_url) : '';
-        const isCurrentPage = fullSlug === cached_url || (fullSlug === 'stories' && cached_url === 'stories/');
+      {links.map(({ _uid, label, link }) => {
+        const processedCachedUrl = sbStripSlugURL(link?.cached_url || '');
+        const processedFullSlug = sbStripSlugURL(fullSlug);
+        const isCurrentPage = processedFullSlug === processedCachedUrl;
+
```
After I updated the sbStripSlugURL function, the current page highlight in the story filter page nav stop working, so I made some more tweaks to the function, and process both the full slug and the cached_url so that they're 100% in the same format before comparing them

sherakama replied:
Aye, the forward and leading slashes get handled this way. 👍

yvonnetangsu commented:
@sherakama This should be working as expected now. Please retest the following especially:

1. Click some links on the page
2. Click some taxonomy chips
3. While on the story filter page, check that the current page nav pill in the nav is highlighted red
4. Click the "All" story filter menu nav item and check that it gets highlighted once it lands on the page
5. Check the sitemap.xml again
6. Check the canonical URLs
7. Look at the code updates

sherakama commented:
> @sherakama I just found something weird, but maybe you know this already? I guess the `cached_url` really is cached, but for some reason I remember it changing when a story is moved in the past 🤔
> 
> I did an experiment and went to a test page I added a long time ago before we did the directory split. There's an existing CTA to the changemaker page (on the left). I added a new one on the right that links to the same changemaker page. See how they have different `cached_url` - the old one is still changemaker but the new one has `cached_url` momentum/changemaker 👀 I've already opened up the old CTA on the left to jiggle it to make sure it has updated itself before I saved and looked at the draft JSON.
> 
> ![Screenshot 2024-10-17 at 10 48 40 AM](https://private-user-images.githubusercontent.com/42749717/377579174-4266e722-2f0d-4db8-a878-708322f6c591.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkyNzE5MTQsIm5iZiI6MTcyOTI3MTYxNCwicGF0aCI6Ii80Mjc0OTcxNy8zNzc1NzkxNzQtNDI2NmU3MjItMmYwZC00ZGI4LWE4NzgtNzA4MzIyZjZjNTkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE4VDE3MTMzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0N2RiNjEzZjIxNWFkMTJiZTA3ZWRiZDlhZTViN2Q5YmRlZGMxN2M2ZTQxYTYzMWFmYzY2NzA2MThiZTZjN2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zr518yCFwYxUXqmLdx7Kyce7hFOaruSMAtq5-YLhxus) ![Screenshot 2024-10-17 at 10 48 33 AM](https://private-user-images.githubusercontent.com/42749717/377579190-2d34a1e1-18d3-4256-8b7a-92855b3a8562.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkyNzE5MTQsIm5iZiI6MTcyOTI3MTYxNCwicGF0aCI6Ii80Mjc0OTcxNy8zNzc1NzkxOTAtMmQzNGExZTEtMThkMy00MjU2LThiN2EtOTI4NTViM2E4NTYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE4VDE3MTMzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE1NjU0YTAzY2Y1Mzg1ZmRjMWNmODY0NWM3ZjZhMjQzZDc3MWRmYWZiMzliNGNlODJmYWVjMGRlNTBkMGZhYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4R0Wc0R_GRHFNEbS-dH4MAPA2_a-N09BUU99ywX8Irc)

Huh, yeah. 

This is in alignment of Storyblok's 'Dont touch the content' when config changes approach. I would have expected the URL to have changed to the referenced story however. 

This case was supposed to have been mitigated by the URL strip function and the checking for the slug first before lopping off the first part. 

Oh man.

sherakama commented on utilities/data/getStoryblokRedirects.mjs:
```diff
@@ -49,7 +49,7 @@ const sanitizeRedirectCode = (code) => {
  */
 const sanitizeSourcePath = (str) => {
-  const parts = str.split('?');
+  const parts = str?.split('?');
```
If `str` isn't a string, it shouldn't have been passed in. I feel that the optional chaining operator should not be here.

yvonnetangsu replied:
Yeah that was while I was troubleshooting while I first got the built error because of the empty redirects. The superficial error points to the split so this was my first attempt. I think we don't need these now - I'll remove.

sherakama commented on utilities/data/getStoryblokRedirects.mjs:
```diff
@@ -100,12 +100,13 @@ export const getStoryblokRedirects = async () => {
     has: extractQueryParameters(entry.content.from),
     destination: entry.content.to,
-    statusCode: sanitizeRedirectCode(entry.content.statusCode),
+    statusCode: sanitizeRedirectCode(entry.content.statusCode || 301),
```
You shouldn't need the `|| 301` here as the sanitize function should default to that already
https://github.com/SU-SWS/ood-giving-campaign/blob/003d5de75eae1ad8d58438e373728b65489d7dc5/utilities/data/getStoryblokRedirects.mjs#L43

sherakama reviewed (approved):
A couple of pendantic code style comments and then you're GTG. Thanks for this work.

yvonnetangsu commented:
Thank you for the review @sherakama. I'm going to merge my annotated image PR into `dev` as well (just style and responsive image updates and Alexis has reviewed). I think I have a 3rd PR for some Momentum updates - after that I'll cut a release 🤠
